mu-plugins/blocks/favorite-button/index.php

```diff
@@ -0,0 +1,151 @@
+ * Get the block settings, set up the filters.
+ */
+function get_block_settings( $post_id ) {
+	/**
+	 * Get the settings for the favorite button, used in rendering and saving favorites.
+	 *
+	 * @param array $settings {
+	 *     Array of settings for the favorite button.
+	 *
+	 *     The return value should use the following format.
+	 *
+	 *     @type callable $add_callback    Callback function to handle adding the current
+	 *                                   item to a user's favorites. The function will
+	 *                                   be passed the post ID and request object. It
+	 *                                   should return a WP_Error, true, or the updated
+	 *                                   count of favorite items.
+	 *     @type callable $delete_callback Callback function to handle removing the current
+	 *                                   item from a user's favorites. Same arguments and
+	 *                                   return value as the `add_callback`.
+	 *     @type int      $count           Number of times this item has been favorited.
+	 *     @type bool     $is_favorite     Check if the current item is favorited.
+	 * }
+	 * @param int $post_id The current post ID.
+	 */
+	$settings = apply_filters( 'wporg_favorite_button_settings', array(), $post_id );
+	if ( empty( $settings ) ) {
+		return false;
+	}
+
+	$settings = wp_parse_args(
+		$settings,
+		array(
+			'add_callback' => '__return_false',
+			'delete_callback' => '__return_false',
+			'count' => 0,
+			'is_favorite' => false,
+		)
+	);
+
+	return $settings;
+}
+
+/**
+ * Initialize the API endpoints.
+ */
+function api_init() {
+	$namespace = 'wporg/v1';
+	$args = array(
+		'id' => array(
+			'validate_callback' => function( $param, $request, $key ) {
+				return is_numeric( $param );
+			},
+			'required' => true,
+		),
+	);
+	register_rest_route(
+		$namespace,
+		'/favorite',
+		array(
+			'methods' => \WP_REST_Server::CREATABLE,
+			'callback' => __NAMESPACE__ . '\add_favorite',
+			'args' => $args,
+			'permission_callback' => 'is_user_logged_in',
+		)
+	);
+	register_rest_route(
+		$namespace,
+		'/favorite',
+		array(
+			'methods' => \WP_REST_Server::DELETABLE,
+			'callback' => __NAMESPACE__ . '\delete_favorite',
+			'args' => $args,
+			'permission_callback' => 'is_user_logged_in',
+		)
+	);
+}
+
+/**
+ * Set the favorite status for a given item.
+ */
+function add_favorite( $request ) {
+	$id = intval( $request['id'] );
+	$settings = get_block_settings( $id );
+	$result = call_user_func( $settings['add_callback'], $id, $request );
+
+	if ( is_wp_error( $result ) ) {
+		return $result;
+	} else if ( false !== $result ) {
+		if ( is_numeric( $result ) ) {
+			return new \WP_REST_Response( $result, 200 );
+		} else {
+			return new \WP_REST_Response( [ 'success' => true ] );
+		}
+	}
+
+	return new \WP_Error(
+		'favorite-failed',
+		// Users should never see this error, so we can leave it untranslated.
+		'Unable to favorite this item.',
```

Comment:
Will $result included any more info that might be useful to a developer?

Reply:
If `$result` is a WP_Error (maybe-useful error info), it's returned as-is in the conditional above. Really, the only reason anyone would reach this code is if the callbacks aren't implemented… which is probably a good error message for devs. I'll update it 🙂

Reply:
Updated error messaging: https://github.com/WordPress/wporg-mu-plugins/pull/635/commits/0ebb8f2ae573ca3ea674716f2df114677fbccbb0